app/services/unsubscribe_service.rb

```diff
@@ -1,37 +1,52 @@
 module UnsubscribeService
   class << self
     def subscriber!(subscriber, reason)
-      unsubscribe!(subscriber, subscriber.active_subscriptions, reason)
+      ActiveRecord::Base.transaction do
+        unsubscribe_subscriptions!(subscriber.active_subscriptions, reason)
+        unsubscribe_subscriber!(subscriber)
+      end
     end
 
     def subscription!(subscription, reason)
-      unsubscribe!(subscription.subscriber, [subscription], reason)
+      ActiveRecord::Base.transaction do
+        unsubscribe_subscriptions!([subscription], reason)
+        unsubscribe_subscriber!(subscription.subscriber)
+      end
     end
 
     def spam_report!(delivery_attempt)
       subscriber_id = delivery_attempt.email.subscriber_id
       subscriber = Subscriber.find(subscriber_id)
-      unsubscribe!(subscriber, subscriber.active_subscriptions, :marked_as_spam, delivery_attempt.email)
+      ActiveRecord::Base.transaction do
+        unsubscribe_subscriptions!(subscriber.active_subscriptions, :marked_as_spam, delivery_attempt.email)
+        unsubscribe_subscriber!(subscriber)
+      end
     end
 
-  private
-
-    def unsubscribe!(subscriber, subscriptions, reason, email = nil)
+    def subscriber_list!(list, reason)
       ActiveRecord::Base.transaction do
-        subscriptions.each do |subscription|
-          subscription.end(reason: reason)
+        subscriptions = list.subscriptions
+        unsubscribe_subscriptions!(subscriptions, reason)
+        Subscriber.where(subscriptions: subscriptions).each do |subscriber|
```

Comment:
Given this is happening in the request (as far as I can see), I don't know how this will cope with lists with many subscriptions. The largest I know of for the legacy things is ~16000, and I'd worry that doing 16000 queries inside the request won't work.

Not something we necessarily need to solve now, but if I'm right, it might be worth improving soon so that this will work with large lists.

Reply:
You're right - made this call asynchronous